### PR TITLE
Added support for HTTP 429 Retry-After as per Issue #99

### PR DESCRIPTION
**Pull request by mariotoffia**
This was added in DefaultPolicy and DefaultBackoff. The latter will
examine the Retry-After response header (if existant) and do a backoff
for the specified amount of time. Otherwise it will default to default
backoff behaviour. 

This addresses Issue #99.

### CODE REVIEW

**Comment by theckman** [comment]
There's PR #70 which adds a new retry policy function, based on the one in `master`. If that gets merged before this one, we'll need to include this change in that new function.

**Comment by jefferai** [comment]
@mariotoffia  #70 was merged. Additionally the CLA isn't signed. Please update/fix and we can look again!

**Comment by mariotoffia** [comment]
Thanks! :)

I've merged and committed - still there's two test errors The report I could download (Artifacts) do not describe the output of the error, just that it has errors on following methods:
```xml
<testsuites>
<testsuite tests="3" failures="0" time="0.000000" name="github.com/hashicorp/go-retryablehttp">
<properties>
<property name="go.version" value="go1.14.2 linux/amd64"/>
</properties>
<testcase classname="" name="TestMain" time="0.000000">
<failure message="Failed" type="">FAIL github.com/hashicorp/go-retryablehttp 0.007s </failure>
</testcase>
<testcase classname="github.com/hashicorp/go-retryablehttp" name="TestRequest" time="0.000000"/>
<testcase classname="github.com/hashicorp/go-retryablehttp" name="TestFromRequest" time="0.000000"/>
</testsuite>
</testsuites>
```

I don't know why those two functions get's errors since I've not altered those. If I e.g. run TestFromRequest locally it works just fine:
```bash
go test -v -run TestFromRequest
=== RUN   TestFromRequest
--- PASS: TestFromRequest (0.00s)
PASS
ok      github.com/hashicorp/go-retryablehttp   0.004s
```

Cheers,
 Mario

**Comment by mateusrangel** [comment]
Any news on why the test are failing?

**Comment by mateusrangel** [comment]
So @mariotoffia, i am checking your code and already detected the point of panic.

**Comment by mateusrangel** [comment]
locally you are testing only TestFromRequest, i did a pure go test before and after your changes and got a 

[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x6db207]

goroutine 9 [running]:
github.com/hashicorp/go-retryablehttp.DefaultBackoff(0x989680, 0x2faf080, 0x0, 0x0, 0x3)

When i comment your code at:

	if resp.StatusCode == http.StatusTooManyRequests {
		if s, ok := resp.Header["Retry-After"]; ok {
			if sleep, err := strconv.ParseInt(s[0], 10, 32); err == nil {
				return time.Duration(int64(time.Second) * sleep)
			}
		}
	}

every thing runs fine, seems like we have a problem in the header check or in the parsing

**Comment by mariotoffia** [comment]
@mateusrangel That's strange - when I do a
```bash
martoffi@dell:~/progs/github/go-retryablehttp$ go test -v -run TestClient_DefaultBackoff429TooManyRequest
```

I do get 
```bash
=== RUN   TestClient_DefaultBackoff429TooManyRequest
2020/06/17 21:10:10 [DEBUG] GET http://127.0.0.1:40959
--- PASS: TestClient_DefaultBackoff429TooManyRequest (0.00s)
PASS
ok      github.com/hashicorp/go-retryablehttp   0.005s
```

When I run it locally. However, I was a bit curious how to get the test output from your CI tool - how do I do that?

Cheers,
 Mario

**Comment by mariotoffia** [comment]
@mateusrangel Ok, I've got it - the
```golang
resp *http.Response
```
is nil. I'm committing a fix now.

**Comment by mateusrangel** [comment]
@mariotoffia Happy to help the debugging :smile:

**Comment by mariotoffia** [comment]
@jefferai is it good to go?

File: client.go [hunk 1]
```diff
@@ -481,7 +489,21 @@ func ErrorPropagatedRetryPolicy(ctx context.Context, resp *http.Response, err er
 // DefaultBackoff provides a default callback for Client.Backoff which
 // will perform exponential backoff based on the attempt number and limited
 // by the provided minimum and maximum durations.
+//
+// It also tries to parse Retry-After response header when a http.StatusTooManyRequests
+// (HTTP Code 429) is found in the resp parameter. Hence it will return the number of
+// seconds the server states it may be ready to process more requests from this client.
 func DefaultBackoff(min, max time.Duration, attemptNum int, resp *http.Response) time.Duration {
+	if nil != resp {
```
Comment (jefferai) [inline client.go:497]:
```suggestion
	if resp != nil {
```

File: client.go [hunk 2]
```diff
@@ -481,7 +489,21 @@ func ErrorPropagatedRetryPolicy(ctx context.Context, resp *http.Response, err er
...
+		if resp.StatusCode == http.StatusTooManyRequests {
+			if s, ok := resp.Header["Retry-After"]; ok {
+				if sleep, err := strconv.ParseInt(s[0], 10, 32); err == nil {
```
Comment (jefferai) [inline client.go:500]:
```suggestion
				if sleep, err := strconv.ParseInt(s[0], 10, 64); err == nil {
```

File: client.go [hunk 3]
```diff
@@ -481,7 +489,21 @@ func ErrorPropagatedRetryPolicy(ctx context.Context, resp *http.Response, err er
...
+					return time.Duration(int64(time.Second) * sleep)
```
Comment (jefferai) [inline client.go:501]:
```suggestion
					return time.Second * time.Duration(sleep)
```

File: client_test.go [hunk 4]
```diff
@@ -523,6 +523,38 @@ func TestClient_CheckRetry(t *testing.T) {
...
+	}))
+	defer ts.Close()
+
+	client := NewClient()
+
+	retryAfter := time.Duration(0)
+	retryable := false
+
+	client.CheckRetry = func(_ context.Context, resp *http.Response, err error) (bool, error) {
+		retryable, _ = DefaultRetryPolicy(context.TODO(), resp, err)
```
Comment (jefferai) [inline client_test.go:539]:
```suggestion
		retryable, _ = DefaultRetryPolicy(context.Background(), resp, err)
```

File: client_test.go [hunk 5]
```diff
@@ -523,6 +523,38 @@ func TestClient_CheckRetry(t *testing.T) {
...
+		retryAfter = DefaultBackoff(client.RetryWaitMin, client.RetryWaitMax, 1, resp)
+		return false, nil
+	}
+
+	_, err := client.Get(ts.URL)
+	if err != nil {
+		t.Fatalf("expected no errors since retryable")
+	}
+
+	if true != retryable {
```
Comment (jefferai) [inline client_test.go:549]:
```suggestion
	if !retryable {
```

File: client_test.go [hunk 6]
```diff
@@ -523,6 +523,38 @@ func TestClient_CheckRetry(t *testing.T) {
...
+func TestClient_DefaultBackoff429TooManyRequest(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Retry-After", "2")
+		http.Error(w, "test_429_body", http.StatusTooManyRequests)
+	}))
+	defer ts.Close()
+
+	client := NewClient()
+
+	retryAfter := time.Duration(0)
```
Comment (jefferai) [inline client_test.go:535]:
```suggestion
	var retryAfter time.Duration
```

File: client_test.go [hunk 7]
```diff
@@ -523,6 +523,38 @@ func TestClient_CheckRetry(t *testing.T) {
...
+	_, err := client.Get(ts.URL)
+	if err != nil {
+		t.Fatalf("expected no errors since retryable")
+	}
+
+	if true != retryable {
+		t.Fatal("Since 429 is recoverable, the default policy shall return true")
+	}
+
+	if time.Second*2 != retryAfter {
```
Comment (jefferai) [inline client_test.go:553]:
```suggestion
	if retryAfter != 2*time.Second {
```

**Comment by mariotoffia** [comment]
@jefferai committed your review points - I'd guess this PR needs to be squashed since it is insane amount of commits :)

File: client.go [hunk 8]
```diff
@@ -425,6 +426,13 @@ func DefaultRetryPolicy(ctx context.Context, resp *http.Response, err error) (bo
 		return true, nil
 	}
 
+	// 429 Too Many Request is recoverable. Sometimes the server puts
```
Comment (jefferai) [inline client.go:429]:
```suggestion
	// 429 Too Many Requests is recoverable. Sometimes the server puts
```

**Pull request merged**